### PR TITLE
[move-analyzer] basic function name completion

### PR DESCRIPTION
**Pull request by 0xYao**
## Motivation

The current move-analyzer has limited code completion support, this PR adds primitive types and function name completion.

Note: the `IdentType` in `FunctionIdentTypeMap` is not used in this PR. However, it will be used for function signature help in a future PR.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/move-language/move/blob/main/CONTRIBUTING.md#developer-workflow)?

Yes

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

- Open the extension in debug mode
- There should be a code completion suggestion for primitive types if you start typing a prefix of any primitive types (e.g. u64, bool etc.)
- There should be a code completion suggestion for any function name prefix inside the same file

## PRs to add after the current one

- Add more intellisense support. (e.g. show function signature help, module method completion, auto imports etc.) Each of these is likely going to be a separate PR for easier review.

### CODE REVIEW

File: language/move-analyzer/src/symbols.rs [hunk 1]
```diff
@@ -214,6 +218,8 @@ pub struct Symbols {
     file_use_defs: BTreeMap<PathBuf, UseDefMap>,
     /// A mapping from file hashes to file names
     file_name_mapping: BTreeMap<FileHash, Symbol>,
+    /// A mapping from filePath to function definitions
+    pub file_functions: BTreeMap<PathBuf, FunctionUseDefMap>,
```
Comment (awelc) [inline language/move-analyzer/src/symbols.rs:222]: Instead of making it public, perhaps you'd consider implementing the right accessor functions for the `Symbols` type? After all you already pass a read-only reference to `Symbols` to `on_completion_request`.

**Review by awelc — CHANGES_REQUESTED** [review]
I am not sure what the intent here is with respect to function name auto-completion as the PR description only says:
```
There should be a code completion suggestion for any function name prefix inside the same file
```
As I understand it, it means that what we want here is auto-completion for functions defined in the same file. If this is correct, then using `UseDef` to recognize these is an overkill, particularly that `UseDef` represents a use of a function paired with information where definition of this function is. Consider the following code:
```
fun some_fun() {}

fun foo() { some_fun(); }

fun bar() { some_fun(); }
```
In the code above you have only one definition of `some_fun` but two `UseDef`s for this function.

More generally, you don't seem to be using any information in the `UseDef` and `FunctionUseDefMap` is used more like a set. Perhaps it makes sense to use the set structure instead (unless I am missing something).

Finally, could you please add a few integration tests in a similar style as #328  does?

**Comment by 0xYao** [comment]
> In the code above you have only one definition of `some_fun` but two `UseDef`s for this function.
> 
> More generally, you don't seem to be using any information in the `UseDef` and `FunctionUseDefMap` is used more like a set. Perhaps it makes sense to use the set structure instead (unless I am missing something).
> 
> Finally, could you please add a few integration tests in a similar style as #328 does?

Correct, I am not using any information in `UseDef` in this PR. The idea is in a following PR, I would add signature help for each function, so I will need the function's signature.

Perhaps storing the `IndentType` is more semantically correct than `UseDef`.

Yep, I will add some integration tests.

**Comment by awelc** [comment]
> > In the code above you have only one definition of `some_fun` but two `UseDef`s for this function.
> > More generally, you don't seem to be using any information in the `UseDef` and `FunctionUseDefMap` is used more like a set. Perhaps it makes sense to use the set structure instead (unless I am missing something).
> > Finally, could you please add a few integration tests in a similar style as #328 does?
> 
> Correct, I am not using any information in `UseDef` in this PR. The idea is in a following PR, I would add signature help for each function, so I will need the function's signature.
> 
> Perhaps storing the `IndentType` is more semantically correct than `UseDef`.

It may be, particularly since you are introducing a new map. If there is future use of the data you are not yet using, please at least mention this in comments to avoid confusion.

> Yep, I will add some integration tests.

Thank you!

**Comment by 0xYao** [comment]
> It may be, particularly since you are introducing a new map. If there is future use of the data you are not yet using, please at least mention this in comments to avoid confusion.

Apologies, will do so next time!

**Comment by 0xYao** [comment]
@awelc added all the suggested changes!

File: language/move-analyzer/editors/code/src/commands/lsp_command.ts [hunk 2]
```diff
@@ -1,24 +1,42 @@
 import type {
```
Comment (awelc) [inline language/move-analyzer/editors/code/src/commands/lsp_command.ts:1]: Could you please conform to the existing indentation scheme to avoid unnecessary changes?

File: language/move-analyzer/editors/code/tests/lsp.test.ts [hunk 3]
```diff
@@ -49,4 +57,43 @@ Mocha.suite('LSP', () => {
         assert.deepStrictEqual(syms[0]?.children[3].name, 'this_is_a_test');
         assert.deepStrictEqual(syms[0]?.children[3]?.detail, '["test", "expected_failure"]');
     });
+
+    Mocha.test('textDocument/completion', async () => {
```
Comment (awelc) [inline language/move-analyzer/editors/code/tests/lsp.test.ts:61]: Could you please include a test for primitive type auto-completion?

**Review by awelc — CHANGES_REQUESTED** [review]
Two small(ish) requests but we are almost ready to go!

File: language/move-analyzer/editors/code/tests/lsp.test.ts (hunk 3 above)
Comment (0xYao) [reply language/move-analyzer/editors/code/tests/lsp.test.ts:70]: done!

File: language/move-analyzer/editors/code/src/commands/lsp_command.ts (hunk 2 above)
Comment (0xYao) [reply language/move-analyzer/editors/code/src/commands/lsp_command.ts]: done

**Comment by 0xYao** [comment]
hey @awelc, do I need to create a new PR because my current PR has merge conflicts?

**Comment by awelc** [comment]
> hey @awelc, do I need to create a new PR because my current PR has merge conflicts?

Please don't. You need to rebase on top of the main branch tip, resolve conflicts, and force push your changes (that's the typical flow used with GitHub).

**Comment by 0xYao** [comment]
> > hey @awelc, do I need to create a new PR because my current PR has merge conflicts?
> 
> Please don't. You need to rebase on top of the main branch tip, resolve conflicts, and force push your changes (that's the typical flow used with GitHub).

I asked, because in `CONTRIBUTING.md`, it says `PRs with merge commits will be rejected.`

**Comment by 0xYao** [comment]
> > hey @awelc, do I need to create a new PR because my current PR has merge conflicts?
> 
> Please don't. You need to rebase on top of the main branch tip, resolve conflicts, and force push your changes (that's the typical flow used with GitHub).

awesome, I misread the `CONTRIBUTING.md` doc

**Comment by 0xYao** [comment]
image

Running `cargo test` in `language/move-prover/move-docgen` throws this error, any idea how to fix it? I didn't change anything in the `language/move-prover/move-docgen` folder.

**Comment by 0xYao** [comment]
@awelc should be good to go now

File: language/move-analyzer/editors/code/tests/lsp.test.ts [hunk 4]
```diff
@@ -126,6 +143,76 @@ Mocha.suite('LSP', () => {
...
+
+        const itemsOnColon = await vscode.commands.executeCommand<Array<vscode.CompletionItem>>(
+            'move-analyzer.textDocumentCompletion',
+            colonParams,
+        );
+
+        assert.ok(itemsOnColon);
+
+        const keywordsOnColon = itemsOnColon.filter(i => i.kind === CompletionItemKind.Keyword);
+        // Primitive types are the only keywords returned after the inserting the colon
```
Comment (awelc) [inline language/move-analyzer/editors/code/tests/lsp.test.ts:210]:
```suggestion
        // Primitive types are the only keywords returned after inserting the colon
```

**Review by awelc — CHANGES_REQUESTED** [review]
Looks great! Just a quick typo fix please before I merge.

**Comment by 0xYao** [comment]
> Looks great! Just a quick typo fix please before I merge.

awesome, done!

**Review by awelc — APPROVED** [review]

File: language/move-analyzer/src/symbols.rs [hunk 5]
```diff
@@ -218,6 +222,8 @@ pub struct Symbols {
     file_use_defs: BTreeMap<PathBuf, UseDefMap>,
     /// A mapping from file hashes to file names
     file_name_mapping: BTreeMap<FileHash, Symbol>,
+    /// A mapping from filePath to function definitions
+    file_functions: BTreeMap<PathBuf, FunctionIdentTypeMap>,
```
Comment (awelc) [inline language/move-analyzer/src/symbols.rs:226]: There is a problem here that needs addressing. We can have more than one module per file and each module may define a function with the same name. This essentially maps file paths to function names which is going to to be a problem if more than one function with the same name is defined in the same file.
